### DISCUSSION TITLE
this ch-run does not support internal SquashFS mounts

### DISCUSSION THREAD

**Discussion by VladimirFokow** [hpc/charliecloud#1851] [General]
Hello! 

I'm following the tutorial here: https://hpc.github.io/charliecloud/tutorial.html#build-from-dockerfile
and getting an error:
```
 ch-run[192857]: error: this ch-run does not support internal SquashFS mounts (ch-run.c:213)
```

I've tested this with such charliecloud versions: 
- 0.32 (is in [ch-run.c:188](https://github.com/hpc/charliecloud/blob/v0.32/bin/ch-run.c#L188)) 
- 0.37 (is in [ch-run.c:213](https://github.com/hpc/charliecloud/blob/master/bin/ch-run.c#L213))

<br />

<details>
<summary>Another (custom) example:</summary>

## Dockerfile:

```
FROM ubuntu:22.04
```

## Commands:

```bash
ch-image build -t test .
ch-image list
```
> ```
> test
> ubuntu:22.04
> ```

then:
```bash
ch-convert test test.sqfs
ls
```
> ```
> Dockerfile  test.sqfs
> ```

then:
```bash
ch-run test.sqfs -- bash
```
> ```
> ch-run[192857]: error: this ch-run does not support internal SquashFS mounts (ch-run.c:213)
> ```

</details>

<br /><br />

After running `./configure`, I noticed this in its output:
```
Building Charliecloud
~~~~~~~~~~~~~~~~~~~~~

. . .

  ch-run(1) internal SquashFS mounting: no
    enabled ... yes
    libfuse3 ... no 
    libsquashfuse_ll ... n/a
    ll.h header ... n/a

. . .

Test suite
~~~~~~~~~~
.
.
.
  recommended tests, squash-unpack mode: no
    basic tests ... yes
    more tests ... no
    pack/unpack SquashFS images ... yes

  recommended tests, squash-mount mode: no
    recommended, squash-unpack mode: no
    internal SquashFS mounting ... no
```

How to solve it? What does it mean?

**Comment by reidpr** [comment]
Hello Vladimir and welcome.

To run SquashFS images, you need to link Charliecloud with libsquashfuse at built time, i.e. you’ll need “internal SquashFS mounting ... yes” from configure. With most distros this requires building and installing SquashFUSE from source.

An alternative is to do the tutorial with directory format images. For example, you’d say something like `ch-convert test /var/tmp/test` and then `ch-run /var/tmp/test -- bash`.

Does that help?

**Reply by VladimirFokow** [reply]
thank you! The reason became clear.

I've tried the directory images and wanted to explore the `.sqfs` format now :)

**Reply by reidpr** [reply]
You should be able to install SquashFUSE without root. You’d do something like `./configure --prefix=$HOME/opt` for SquashFUSE and then `./configure --with-libsquashfuse=$HOME/opt` for Charliecloud.
